IdentityCore/src/webview/MSIDWebviewAuthorization.m

```diff
@@ -26,34 +26,176 @@
 {
-    return nil;
+    if (![self setCurrentWebSession:session])
+    {
+        NSError *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInteractiveSessionAlreadyRunning, @"Only one interactive session is allowed at a time.", nil, nil, nil, context.correlationId, nil);
+        completionHandler(nil, error);
+        return;
+    }
+    
+    void (^startCompletionBlock)(NSURL *, NSError *) = ^void(NSURL *callbackURL, NSError *error) {
+        if (error) {
+            completionHandler(nil, error);
+            [MSIDWebviewAuthorization clearCurrentWebAuthSessionAndFactory];
+            return;
+        }
+        
+        NSError *responseError = nil;
+        
+        MSIDWebviewResponse *response = [s_currentSession.factory responseWithURL:callbackURL
+                                                                     requestState:s_currentSession.requestState
+                                                                      verifyState:s_currentSession.verifyState
+                                                                          context:nil
+                                                                            error:&responseError];
+        
+        completionHandler(response, responseError);
+        [MSIDWebviewAuthorization clearCurrentWebAuthSessionAndFactory];
+    };
+    
+    [s_currentSession.webviewController startWithCompletionHandler:startCompletionBlock];
 }
 
-+ (BOOL)handleURLResponse:(NSURL *)url
+
++ (BOOL)setCurrentWebSession:(MSIDWebviewSession *)session
+
 {
+    @synchronized([MSIDWebviewAuthorization class])
+    {
+        if (s_currentSession) {
+            MSID_LOG_INFO(nil, @"Session is already running. Please wait or cancel the session before setting it new.");
+            return NO;
+        }
+        
+        s_currentSession = session;
+        
+        return YES;
+    }
     return NO;
 }
 
+
++ (void)clearCurrentWebAuthSessionAndFactory
+{
+    @synchronized ([MSIDWebviewAuthorization class])
+    {
+        if (!s_currentSession)
+        {
+            // There's no error param because this isn't on a critical path. Just log that you are
+            // trying to clear a session when there isn't one.
+            MSID_LOG_INFO(nil, @"Trying to clear out an empty session");
+        }
+        
+        s_currentSession = nil;
+    }
+}
+
++ (MSIDWebviewSession *)currentSession
+{
+    return s_currentSession;
+}
+
++ (void)cancelCurrentSession
+{
+    @synchronized([MSIDWebviewAuthorization class])
+    {
+        if (s_currentSession)
+        {
+            [s_currentSession.webviewController cancel];
+            s_currentSession = nil;
+        }
+    }
+}
+
+
++ (BOOL)handleURLResponseForSystemWebviewController:(NSURL *)url;
+{
+#if TARGET_OS_IPHONE
+    @synchronized([MSIDWebviewAuthorization class])
+    {
+        if (s_currentSession &&
+            [(NSObject *)s_currentSession.webviewController isKindOfClass:MSIDSystemWebviewController.class])
```

Comment:
nit: do you need to cast to NSObject?

Reply:
yeah because .webviewController is an id<..>.